File: problem_0053.py
#!/usr/bin/env python3
"""
There are exactly ten ways of selecting three from five, 12345:
123, 124, 125, 134, 135, 145, 234, 235, 245, and 345
In combinatorics, we use the notation, 5C3 = 10.
In general,
nCr = n!/r!(n−r)!, where r ≤ n, n! = n×(n−1)×...×3×2×1, and 0! = 1.
It is not until n = 23, that a value exceeds one-million: 23C10 = 1144066.
How many, not necessarily distinct, values of  nCr, for 1 ≤ n ≤ 100, are greater than one-million?
"""

from time import time
import logging
import sys

logging.basicConfig(
    stream=sys.stderr,
    level=logging.INFO,
    format='%(levelname)s:%(message)s')

def main():
    count = 0
    for n in range(1,101):
        for r in range(n):
            combinations = fact(n)/(fact(r)*fact(n-r))
            if combinations > 1e6:
                logging.debug(
                    "There are {} ways of selection {} from {}".format(
                        combinations, r, n))
                count += 1
    print('Answer: {}'.format(count))

def fact(num):
    product = 1
    for i in range(1,num+1):
        product *= i
    return product

if __name__ == "__main__":
    start = time()
    main()
    print("That took {:4.2f} seconds".format(time() - start))

File: problem_0053.hs
{-# LANGUAGE DeriveDataTypeable, RecordWildCards #-}

import Test.HUnit ((@=?), assertBool, runTestTT, Test(..))
import Text.Printf (printf)
import System.Console.CmdArgs
import Data.Time (getCurrentTime, diffUTCTime)

import Data.List (nub)

-- There are exactly ten ways of selecting three from five, 12345:
-- 123, 124, 125, 134, 135, 145, 234, 235, 245, and 345
-- In combinatorics, we use the notation, 5C3 = 10.
-- In general,
-- nCr = n!/r!(n−r)!, where r ≤ n, n! = n×(n−1)×...×3×2×1, and 0! = 1.
-- It is not until n = 23, that a value exceeds one-million: 23C10 = 1144066.
-- How many, not necessarily distinct, values of nCr, for 1 ≤ n ≤ 100, are greater than one-million?

data EulerArgs =
    Euler
    | AdHoc{ n::Integer, limit::Integer }
    | UnitTest
    deriving (Show, Data, Typeable)

problem0053 :: Integer -> Integer -> [(Integer,Integer,Integer)]
problem0053 limit bigNum = [(n, r, nCr) | n <- [1..limit], r <- [0..n], 
    let nCr = c n r, nCr > bigNum]

c :: Integer -> Integer -> Integer
c n r = div (fact n) $ (fact r) * (fact (n - r))
    where
        fact = (!!) factorials . fromIntegral

cTest = [
    1144066 @=? c 23 10,
    10 @=? c 5 3]

factorials :: [Integer]
factorials = scanl (*) 1 [1..]

factorialsTest = [
    2 @=? (factorials !! 2) ,
    1 @=? (factorials !! 1) ,
    1 @=? (factorials !! 0)]

unitTests = map TestCase $
    factorialsTest ++
    cTest

exec :: EulerArgs -> IO ()
exec Euler = do
    let  answer = problem0053 100 (10 ^ 6)
    printf "Answer: %d\n" (length answer)
exec AdHoc{..} = do
    let answer = problem0053 n (10 ^ 6)
    mapM_ (\(n, r, nCr) -> printf "%d C %d = %d\n" n r nCr) answer
exec UnitTest = do
    runTestTT $ TestList unitTests
    return ()

main :: IO ()
main = do
    args <- cmdArgs $ modes [
        Euler,
        AdHoc{ n=23, limit=(10^6)},
        UnitTest]
    start <- getCurrentTime
    exec args
    stop <- getCurrentTime
    printf "That took %s\n"  $ show $ diffUTCTime stop start
